src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -409,7 +410,7 @@ private TypeSymbol NarrowDiscriminatedObjectType(TypeValidatorConfig config, Obj
                                     return x.PropertyStringLiteralMismatchWithSuggestion(shouldWarn, targetType.DiscriminatorKey, targetType.DiscriminatorKeysUnionType, stringLiteralDiscriminator.Name, suggestion);
                                 }
 
-                                return x.PropertyTypeMismatch(shouldWarn, sourceDeclaration, targetType.DiscriminatorKey, targetType.DiscriminatorKeysUnionType, discriminatorType);
+                                return x.PropertyTypeMismatch(shouldWarn, sourceDeclaration, targetType.DiscriminatorKey, targetType.DiscriminatorKeysUnionType, discriminatorType, config.IsResourceDeclaration && !targetType.DiscriminatorProperty.Flags.HasFlag(TypePropertyFlags.SystemProperty));
```

Comment:
>config.IsResourceDeclaration && !targetType.DiscriminatorProperty.Flags.HasFlag(TypePropertyFlags.SystemProperty) [](http://example.com/codeflow?start=180&length=113)

L388 includes this condition. Can you replace with a new variable so the relationship is more obvious?

Reply:
I could but the PR is merged... I'll refactor this in some next PR

Reply:
However, on the other side I think we'd need to refactor this warning/error mechanism as it feels to be fix on a fix to a fix.